File: java/src/main/java/sk/epostak/sdk/resources/PayloadsResource.java
package sk.epostak.sdk.resources;

import sk.epostak.sdk.HttpClient;
import sk.epostak.sdk.models.BatchExtractResult;
import sk.epostak.sdk.models.ConvertResult;
import sk.epostak.sdk.models.ExtractResult;
import sk.epostak.sdk.models.ParsedInvoice;
import sk.epostak.sdk.models.ValidationResult;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Payload Assistant helpers for OCR, parse, convert, and validation. */
public final class PayloadsResource {

    private final HttpClient http;

    public PayloadsResource(HttpClient http) {
        this.http = http;
    }

    public ExtractResult extract(byte[] fileBytes, String fileName, String mimeType) {
        return http.postMultipart("/payloads/extract", fileBytes, fileName, mimeType, ExtractResult.class);
    }

    public BatchExtractResult extractBatch(List<ExtractResource.FileInput> files) {
        List<HttpClient.FileUpload> uploads = files.stream()
                .map(f -> new HttpClient.FileUpload(f.data(), f.fileName(), f.mimeType()))
                .toList();
        return http.postMultipartBatch("/payloads/extract/batch", uploads, BatchExtractResult.class);
    }

    public ParsedInvoice parse(String xml) {
        return http.post("/payloads/parse", Map.of("xml", xml), ParsedInvoice.class);
    }

    public ConvertResult convert(String inputFormat, String outputFormat, Object document) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("input_format", inputFormat);
        body.put("output_format", outputFormat);
        body.put("document", document);
        return http.post("/payloads/convert", body, ConvertResult.class);
    }

    public ValidationResult validate(Object request) {
        return http.post("/payloads/validate", request, ValidationResult.class);
    }
}

File: dotnet/src/EPostak/Resources/PayloadsResource.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using EPostak.Models;

namespace EPostak.Resources;

/// <summary>Payload Assistant helpers for OCR, parse, convert, and validation.</summary>
public sealed class PayloadsResource
{
    private readonly HttpRequestor _http;

    internal PayloadsResource(HttpRequestor http) => _http = http;

    public Task<ExtractResult> ExtractAsync(Stream stream, string mimeType, string? fileName = null, CancellationToken ct = default)
        => ExtractMultipartAsync(stream, mimeType, fileName, fields: null, ct);

    /// <summary>Extract one file while applying caller-reviewed OCR corrections.</summary>
    /// <param name="fields">Corrected values serialized into the multipart <c>fields</c> JSON part.</param>
    public Task<ExtractResult> ExtractWithFieldsAsync(
        Stream stream,
        string mimeType,
        IReadOnlyDictionary<string, object?>? fields,
        string? fileName = null,
        CancellationToken ct = default)
        => ExtractMultipartAsync(stream, mimeType, fileName, fields, ct);

    private Task<ExtractResult> ExtractMultipartAsync(
        Stream stream,
        string mimeType,
        string? fileName,
        IReadOnlyDictionary<string, object?>? fields,
        CancellationToken ct)
    {
        var content = new MultipartFormDataContent();
        var streamContent = new StreamContent(stream);
        streamContent.Headers.ContentType = new MediaTypeHeaderValue(mimeType);
        content.Add(streamContent, "file", fileName ?? "document");
        if (fields is not null)
        {
            content.Add(
                new StringContent(
                    JsonSerializer.Serialize(fields, HttpRequestor.JsonOptions),
                    Encoding.UTF8,
                    "application/json"),
                "fields");
        }
        return _http.RequestMultipartAsync<ExtractResult>(HttpMethod.Post, "/payloads/extract", content, ct);
    }

    public Task<BatchExtractResult> ExtractBatchAsync(IEnumerable<ExtractFile> files, CancellationToken ct = default)
    {
        var content = new MultipartFormDataContent();
        foreach (var file in files)
        {
            var streamContent = new StreamContent(file.Stream);
            streamContent.Headers.ContentType = new MediaTypeHeaderValue(file.MimeType);
            content.Add(streamContent, "files", file.FileName ?? "document");
        }
        return _http.RequestMultipartAsync<BatchExtractResult>(HttpMethod.Post, "/payloads/extract/batch", content, ct);
    }

    public Task<ParsedInvoice> ParseAsync(string xml, CancellationToken ct = default)
        => _http.RequestAsync<ParsedInvoice>(HttpMethod.Post, "/payloads/parse", new { xml }, ct);

    public Task<ConvertResult> ConvertAsync(ConvertRequest request, CancellationToken ct = default)
        => _http.RequestAsync<ConvertResult>(HttpMethod.Post, "/payloads/convert", request, ct);

    public Task<ValidationResult> ValidateAsync(object request, CancellationToken ct = default)
        => _http.RequestAsync<ValidationResult>(HttpMethod.Post, "/payloads/validate", request, ct);
}
